profiler/test/Datadog.Profiler.IntegrationTests/Helpers/MockDatadogAgent.cs

```diff
@@ -4,142 +4,344 @@
+
+                var buffer = new byte[1 << 32];
+
+                // _nbTime == 1 ==> no data. This can be the native part checking the existence of the pipe
+                if (_nbTime != 1)
+                {
+                    await ss.ReadAsync(buffer, cancellationToken);
+                    await ss.WriteAsync(_responseBytes, cancellationToken);
+                    NbCallsOnProfilingEndpoint++;
                 }
-                catch (HttpListenerException x)
+            }
+        }
+
+        internal static class DatadogHttpValues
+        {
+            public const char CarriageReturn = '\r';
+            public const char LineFeed = '\n';
+            public const string CrLf = "\r\n";
+        }
+
+        internal class PipeServer : IDisposable
+        {
+            private const int ConcurrentInstanceCount = 5;
+            private readonly CancellationTokenSource _cancellationTokenSource;
+            private readonly string _pipeName;
+            private readonly PipeDirection _pipeDirection;
+            private readonly Func<NamedPipeServerStream, CancellationToken, Task> _handleReadFunc;
+            private readonly ConcurrentBag<Task> _tasks = new();
+            private readonly Action<string> _log;
+            private int _instanceCount = 0;
+
+            public PipeServer(
+                string pipeName,
+                PipeDirection pipeDirection,
+                CancellationTokenSource tokenSource,
+                Func<NamedPipeServerStream, CancellationToken, Task> handleReadFunc,
+                Action<string> log)
+            {
+                _cancellationTokenSource = tokenSource;
+                _pipeDirection = pipeDirection;
+                _pipeName = pipeName;
+                _handleReadFunc = handleReadFunc;
+                _log = log;
+            }
+
+            public Task Start()
+            {
+                for (var i = 0; i < ConcurrentInstanceCount; i++)
                 {
-                    // listener was stopped,
-                    // ignore to let the loop end and the method return
-                    message = $"NbReceivedCalls = {NbCallsOnProfilingEndpoint} | HttpListenerException({x.ErrorCode}, {x.NativeErrorCode}): {x.Message}";
+                    _log("Starting PipeServer " + _pipeName);
+                    using var mutex = new ManualResetEventSlim();
+                    var startPipe = StartNamedPipeServer(mutex);
+                    _tasks.Add(startPipe);
+                    mutex.Wait(5_000);
                 }
-                catch (ObjectDisposedException x)
+
+                return Task.CompletedTask;
+            }
+
+            public void Dispose()
+            {
+                _log("Waiting for PipeServer Disposal " + _pipeName);
+                Task.WaitAll(_tasks.ToArray(), TimeSpan.FromSeconds(10));
+            }
+
+            private async Task StartNamedPipeServer(ManualResetEventSlim mutex)
```

Comment:
I would rename mutex into stopEvent because here it is used as an event to stop the processing in the caller. A mutex does the opposite

Reply:
✔️